sbncode/EventGenerator/MeVPrtl/Tools/NuMiKaonGen_tool.cc

```diff
@@ -0,0 +1,330 @@
+    simb::MCFlux MakeMCFlux(const bsim::Dk2Nu &dk2nu);
+    double LoadPOT();
+
+    // no weights
+    float MaxWeight() override { return -1.; }
+
+private:
+  // config
+  std::string fSearchPath;
+  std::vector<std::string> fSearchPatterns;
+  unsigned long fMaxFluxFileMB;
+  std::string fFluxCopyMethod;
+  bool fRandomizeFiles;
+
+  std::string fTreeName;
+  std::string fMetaTreeName;
+
+  // info for tracking files
+  unsigned fFileIndex;
+  bool fNewFile;
+  std::vector<std::string> fFluxFiles;
+
+  // info for tracking entry in file
+  unsigned fEntry;
+  unsigned fEntryStart;
+
+  // ROOT Holders
+  TTree *fFluxTree;
+  TFile *fFluxFile;
+  bsim::Dk2Nu *fDk2Nu;
+
+  // count POT
+  double fAccumulatedPOT;
+  double fThisFilePOT;
+  
+};
+
+NuMiKaonGen::NuMiKaonGen(fhicl::ParameterSet const &pset):
+  IMeVPrtlStage("NuMiKaonGen") 
+{
+  configure(pset);
+
+  // copy the flux files locally
+  fFluxFiles = LoadFluxFiles();
+
+  // setup indices
+  fFileIndex = 0;
+  fEntry = 0;
+  fEntryStart = 0;
+  fNewFile = true;
+  fFluxTree = NULL;
+  fFluxFile = NULL;
+  fDk2Nu = new bsim::Dk2Nu;
+
+  fAccumulatedPOT = 0.;
+  fThisFilePOT = 0.;
+    
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+NuMiKaonGen::~NuMiKaonGen()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+void NuMiKaonGen::configure(fhicl::ParameterSet const &pset)
+{
+  fSearchPath = pset.get<std::string>("SearchPath");
+  fSearchPatterns = pset.get<std::vector<std::string>>("FluxFiles");
+  fMaxFluxFileMB = pset.get<unsigned long>("MaxFluxFileMB", 2 * 1024);
+  fFluxCopyMethod = pset.get<std::string>("FluxCopyMethod", "IFDH");
+  fTreeName = pset.get<std::string>("TreeName");
+  fMetaTreeName = pset.get<std::string>("MetaTreeName");
+  fRandomizeFiles = pset.get<bool>("RandomizeFiles");
+
+  std::cout << "Searching for flux files at path: " << fSearchPath << std::endl;
+  std::cout << "With patterns:\n";
+  for (const std::string &s: fSearchPatterns) std::cout << s << std::endl;
+  std::cout << "With copy method: " << fFluxCopyMethod << std::endl;
+}
+
+std::vector<std::string> NuMiKaonGen::LoadFluxFiles() {
+  art::ServiceHandle<IFDH> ifdhp;
+
+  std::vector<std::pair<std::string, long>> allFiles;
+
+  // find the flux files
+  for (unsigned i = 0; i < fSearchPatterns.size(); i++) {
+    std::vector<std::pair<std::string, long>> thisList = ifdhp->findMatchingFiles(fSearchPath, fSearchPatterns[i]);
+    std::copy (thisList.begin(), thisList.end(), std::back_inserter(allFiles));
+  }
+
+  // first randomize the flux files
+  std::vector<unsigned long> order(allFiles.size(), 0);
+  if (fRandomizeFiles) {
+    std::vector<double> rand(allFiles.size(), 0.);
+    CLHEP::RandFlat::shootArray(fEngine, rand.size(), &rand[0]);
+    TMath::Sort(allFiles.size(), &rand[0], &order[0], false);
```

Comment:
rand.data(), order.data() are a bit more elengant, but this is a minor comment anyway